xgcm/test/test_grid_ufunc.py

```diff
@@ -81,6 +83,106 @@ def test_invalid_signatures(self, signature):
             _parse_grid_ufunc_signature(signature)
 
 
+class TestGetSignatureFromTypeHints:
+    def test_no_args_to_annotate(self):
+        @as_grid_ufunc()
+        def ufunc():
+            ...
+
+        assert str(ufunc.signature) == "()->()"
+
+    def test_annotated_args(self):
+        # TODO test hints without annotations
+        # TODO test hints with annotations that don't conform to Xgcm
+
+        @as_grid_ufunc()
+        def ufunc(a: Gridded[np.ndarray, "X:center"]):
+            ...
+
+        assert str(ufunc.signature) == "(X:center)->()"
+
+        @as_grid_ufunc()
+        def ufunc(a: Gridded[np.ndarray, "X:center,Y:center"]):
+            ...
+
+        assert str(ufunc.signature) == "(X:center,Y:center)->()"
+
+        @as_grid_ufunc()
+        def ufunc(
+            a: Gridded[np.ndarray, "X:left"],
+            b: Gridded[np.ndarray, "Y:right"],
+        ):
+            ...
+
+        assert str(ufunc.signature) == "(X:left),(Y:right)->()"
+
+    def test_annotated_return_args(self):
+        @as_grid_ufunc()
+        def ufunc() -> Gridded[np.ndarray, "X:center"]:
+            ...
+
+        assert str(ufunc.signature) == "()->(X:center)"
+
+        @as_grid_ufunc()
+        def ufunc() -> Gridded[np.ndarray, "X:left,Y:right"]:
+            ...
+
+        assert str(ufunc.signature) == "()->(X:left,Y:right)"
+
+        @as_grid_ufunc()
+        def ufunc() -> Tuple[
+            Gridded[np.ndarray, "X:left"], Gridded[np.ndarray, "Y:right"]
+        ]:
+            ...
+
+        assert str(ufunc.signature) == "()->(X:left),(Y:right)"
+
+        @as_grid_ufunc()
+        def ufunc(
+            a: Gridded[xr.DataArray, "X:center"]
+        ) -> Gridded[np.ndarray, "X:left"]:
+            ...
+
+        assert str(ufunc.signature) == "(X:center)->(X:left)"
+
+    def test_invalid_arg_annotation(self):
+        with pytest.raises(ValueError, match="Not a valid grid ufunc signature"):
+
+            @as_grid_ufunc()
+            def ufunc(a: Gridded[np.ndarray, "X:Mars"]):
+                ...
+
+    def test_invalid_return_arg_annotation(self):
+        with pytest.raises(ValueError, match="Not a valid grid ufunc signature"):
+
+            @as_grid_ufunc()
+            def ufunc() -> Gridded[np.ndarray, "X:Venus"]:
+                ...
+
+    def test_both_sig_kwarg_and_hints_given(self):
+        with pytest.raises(
+            ValueError, match="only one of either type hints or signature kwarg"
+        ):
+
+            @as_grid_ufunc(signature="(X:center)->(X:left)")
+            def ufunc(
+                a: Gridded[np.ndarray, "X:center"]
+            ) -> Gridded[np.ndarray, "X:left"]:
+                ...
+
+    def test_type_hint_as_numpy_ndarray(self):
+        # TODO I want this to fail mypy but it doesn't
+        @as_grid_ufunc()
+        def ufunc1(a: Gridded[str, "X:center"]):
```

Comment:
We should turn this into a MRE and open a mypy issue.

Reply:
Changing `Gridded` to `Annotated` appears to have fixed this.